spider3.py

```diff
@@ -0,0 +1,332 @@
+            self.led_time = self.time
+
+    def set_raw_speed(self, fwd_rev_drive, left_right):
+        print('set_raw_speed', fwd_rev_drive, left_right)
+        self.speed_cmd = [fwd_rev_drive, left_right]
+
+    def update(self):
+        while True:
+            packet = self.can.readPacket()
+            self.check_modules(packet)
+            self.update_status_word(packet)
+            for (name,e) in self.extensions:
+                e(self, packet[0], packet[1])
+            
+            # make sure that all updates get also termination SYNC (0x80)
+            # there is no SYNC on Spider3 yet, use 0x200 for now
+            if packet[0] == 0x200:
+                break
+        
+        # send data related to other sources
+        for (id,fce) in self.data_sources:
+            data = fce()
+            if data != None:
+                for (name,e) in self.extensions:
+                    e(self, id, data)
+
+        self.time += 1.0/self.UPDATE_TIME_FREQUENCY  
+        self.send_speed()
+        self.send_LED()
+
+    def stop(self):
+        "send stop command and make sure robot really stops"
+#        self.canproxy.stop()
+        self.set_raw_speed(0, 0)
+        for i in range(10):
+            self.update()
+            # TODO verify encoders/motion
+
+
+def velodyne_data_extension(robot, id, data):
+    if id=='velodyne':
+        robot.velodyne_data = data
+
+
+def self_test(metalog):
+    assert metalog is not None
+    can_log_name = metalog.getLog('can')
+    if metalog.replay:
+        if metalog.areAssertsEnabled():
+            can = CAN(ReplayLog(can_log_name), skipInit=True)
+        else:
+            can = CAN(ReplayLogInputsOnly(can_log_name), skipInit=True)
+    else:
+        can = CAN()
+        can.relog(can_log_name)
+    #can.resetModules()
+    robot = Spider3(can=can)
+
+    robot.localization = None
+    for sensor_name in ['gps',]:
+        attach_sensor(robot, sensor_name, metalog)
+
+    # move this to helper
+    velodyne_log_name = metalog.getLog('velodyne_dist')
+    print(velodyne_log_name)
+    sensor = Velodyne(metalog=metalog)
+    if metalog.replay:
+        robot.velodyne = DummySensor()
+        function = SourceLogger(None, velodyne_log_name).get
+    else:
+        robot.velodyne = VelodyneThread(sensor)
+        function = SourceLogger(robot.velodyne.scan_safe_dist, velodyne_log_name).get
+    robot.velodyne_data = None
+    robot.register_data_source('velodyne', function, velodyne_data_extension) 
+
+    robot.velodyne.start()  # the data source is active, so it is necessary to read-out data
+
+#    robot.stop()
+#    robot.set_desired_speed(0.5)
+
+    try:
+        robot.extensions.append(('emergency_stop', emergency_stop_extension))
+        run_it(robot)
+    except EmergencyStopException:
+        print("Emergency STOP Exception!")
+
+    robot.extensions = []
+    robot.stop()
+    robot.velodyne.requestStop()
+    robot.wait(3.0)
+
+    
+def run_it(robot):
```

Comment:
I must be confused. Or maybe I am just not up-to-date with the plans for osgar/jd/spider. We are not going to use the launcher pattern for spider? Is CAN going to get its own thread? Or are those on the radar for some later time? Or not at all?

Reply:
Well, this is "on the way" to common launcher. At the moment there are no encoders so there is no way ho to convince Spider3 to move 1m ... and as such it is hard to apply some "general framework". spider3.py is the first step to get it moving.